api/sendfild/service/waitmanager/waitrunner.go

```diff
@@ -0,0 +1,188 @@
+package waitmanager
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/filecoin-project/lotus/api/apistruct"
+	"github.com/ipfs/go-cid"
+	"github.com/textileio/powergate/v2/lotus"
+	"github.com/textileio/textile/v2/api/sendfild/service/store"
+)
+
+type WaitResult struct {
+	LatestMessageCid string
+	Err              error
+}
+
+type WaitRunner struct {
+	messageCid       string
+	confidence       uint64
+	waitTimeout      time.Duration
+	store            *store.Store
+	lotusClient      *apistruct.FullNodeStruct
+	closeLotusClient func()
+	listeners        []chan WaitResult
+	lock             sync.Mutex
+	mainCtx          context.Context
+	mainCtxCancel    context.CancelFunc
+	waitCtx          context.Context
+	waitCtxCancel    context.CancelFunc
+}
+
+func NewWaitRunner(ctx context.Context, messageCid string, confidence uint64, waitTimeout time.Duration, store *store.Store, cb lotus.ClientBuilder) (*WaitRunner, error) {
+	mainCtx, mainCtxCancel := context.WithCancel(ctx)
+
+	w := &WaitRunner{
+		messageCid:    messageCid,
+		confidence:    confidence,
+		waitTimeout:   waitTimeout,
+		store:         store,
+		mainCtx:       mainCtx,
+		mainCtxCancel: mainCtxCancel,
+	}
+
+	var err error
+	w.lotusClient, w.closeLotusClient, err = cb(mainCtx)
+	if err != nil {
+		return nil, fmt.Errorf("creating lotus client: %v", err)
+	}
+
+	return w, nil
+}
+
+func (w *WaitRunner) Start() {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	if w.waitCtxCancel != nil {
+		return
+	}
```

Comment:
`Start` is meant to be a no-op if the runner is already started. We use `w.waitCtxCancel` property as a flag to detect if we're started or not. We synchronize access to that check to be safe when calls from many goroutines are coming in.